File: problem002/2.pl
#!/usr/bin/env perl
use strict;
use warnings;
my @h;

sub fib{
	my $n = shift;
	if (!defined($h[$n])) {
		if ($n < 2) {
			$h[$n] = 1;
			return $h[$n];
		} else {
			$h[$n] = fib($n - 1) + fib($n - 2);
			return $h[$n];
		}
	} else {
		return $h[$n];
	}
}

my $x = 2;
my $total = 0;
my $i = fib($x);
while ($i < 4000000){
	$total += $i;
	$x += 3;
	$i = fib($x);
}

print($total);

File: problem002/2.php
#!/usr/bin/env php
<?php
//slow
$h = Array();

function fib($n, $h){
	if ($h[$n] == null) {
		if ($n < 2) {
			$h[$n] = 1;
			return $h[$n];
		} else {
			$h[$n] = fib($n - 1, $h) + fib($n - 2, $h);
			return $h[$n];
		}
	} else {
		return $h[$n];
	}
}

$total = 0;
$x = 2;
$i = fib($x, $h);
while ($i < 4000000) {
	$total += $i;
	$x += 3;
	$i = fib($x, $h);
}

print($total);
?>
